File: lsp/java/src/main/java/com/itsaky/androidide/lsp/java/JavaLanguageServer.java
/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.itsaky.androidide.lsp.java;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.itsaky.androidide.eventbus.events.editor.DocumentChangeEvent;
import com.itsaky.androidide.eventbus.events.editor.DocumentCloseEvent;
import com.itsaky.androidide.eventbus.events.editor.DocumentOpenEvent;
import com.itsaky.androidide.eventbus.events.editor.DocumentSelectedEvent;
import com.itsaky.androidide.javac.services.fs.CacheFSInfoSingleton;
import com.itsaky.androidide.javac.services.fs.CachingJarFileSystemProvider;
import com.itsaky.androidide.lsp.api.ILanguageClient;
import com.itsaky.androidide.lsp.api.ILanguageServer;
import com.itsaky.androidide.lsp.api.IServerSettings;
import com.itsaky.androidide.lsp.internal.model.CachedCompletion;
import com.itsaky.androidide.lsp.java.actions.JavaCodeActionsMenu;
import com.itsaky.androidide.lsp.java.compiler.JavaCompilerService;
import com.itsaky.androidide.lsp.java.compiler.SourceFileManager;
import com.itsaky.androidide.lsp.java.models.JavaServerSettings;
import com.itsaky.androidide.lsp.java.providers.CodeFormatProvider;
import com.itsaky.androidide.lsp.java.providers.CompletionProvider;
import com.itsaky.androidide.lsp.java.providers.DefinitionProvider;
import com.itsaky.androidide.lsp.java.providers.JavaDiagnosticProvider;
import com.itsaky.androidide.lsp.java.providers.JavaSelectionProvider;
import com.itsaky.androidide.lsp.java.providers.ReferenceProvider;
import com.itsaky.androidide.lsp.java.providers.SignatureProvider;
import com.itsaky.androidide.lsp.java.utils.AnalyzeTimer;
import com.itsaky.androidide.lsp.java.utils.CancelChecker;
import com.itsaky.androidide.lsp.models.CodeFormatResult;
import com.itsaky.androidide.lsp.models.CompletionParams;
import com.itsaky.androidide.lsp.models.CompletionResult;
import com.itsaky.androidide.lsp.models.DefinitionParams;
import com.itsaky.androidide.lsp.models.DefinitionResult;
import com.itsaky.androidide.lsp.models.DiagnosticResult;
import com.itsaky.androidide.lsp.models.ExpandSelectionParams;
import com.itsaky.androidide.lsp.models.FormatCodeParams;
import com.itsaky.androidide.lsp.models.LSPFailure;
import com.itsaky.androidide.lsp.models.ReferenceParams;
import com.itsaky.androidide.lsp.models.ReferenceResult;
import com.itsaky.androidide.lsp.models.SignatureHelp;
import com.itsaky.androidide.lsp.models.SignatureHelpParams;
import com.itsaky.androidide.lsp.util.LSPEditorActions;
import com.itsaky.androidide.models.Range;
import com.itsaky.androidide.projects.FileManager;
import com.itsaky.androidide.projects.ProjectManager;
import com.itsaky.androidide.projects.api.ModuleProject;
import com.itsaky.androidide.projects.api.Project;
import com.itsaky.androidide.utils.DocumentUtils;
import com.itsaky.androidide.utils.VMUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.nio.file.Path;
import java.util.Collections;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;

public class JavaLanguageServer implements ILanguageServer {

  public static final String SERVER_ID = "java";
  private final CompletionProvider completionProvider;
  private final JavaDiagnosticProvider diagnosticProvider;
  private ILanguageClient client;
  private IServerSettings settings;
  private Path selectedFile;
  private final AnalyzeTimer timer = new AnalyzeTimer(this::analyzeSelected);
  private CachedCompletion cachedCompletion;

  public JavaLanguageServer() {
    this.completionProvider = new CompletionProvider();
    this.diagnosticProvider = new JavaDiagnosticProvider();
    this.cachedCompletion = CachedCompletion.EMPTY;

    applySettings(getSettings());

    if (!EventBus.getDefault().isRegistered(this)) {
      EventBus.getDefault().register(this);
    }
  }

  public IServerSettings getSettings() {
    if (settings == null) {
      settings = JavaServerSettings.getInstance();
    }

    return settings;
  }

  @Override
  public String getServerId() {
    return SERVER_ID;
  }

  @Override
  public void shutdown() {
    JavaCompilerProvider.getInstance().destory();
    SourceFileManager.clearCache();
    CacheFSInfoSingleton.INSTANCE.clearCache();
    CachingJarFileSystemProvider.INSTANCE.clearCache();
    EventBus.getDefault().unregister(this);

    timer.cancel();
  }

  @Override
  public void connectClient(@Nullable ILanguageClient client) {
    this.client = client;
  }

  @Nullable
  @Override
  public ILanguageClient getClient() {
    return this.client;
  }

  @Override
  public void applySettings(@Nullable IServerSettings settings) {
    this.settings = settings;
  }

  @Override
  public void setupWithProject(@NonNull final Project project) {

    LSPEditorActions.ensureActionsMenuRegistered(JavaCodeActionsMenu.class);

    // Once we have project initialized
    // Destory the NO_MODULE_COMPILER instance
    JavaCompilerService.NO_MODULE_COMPILER.destroy();

    // Clear cached file managers
    SourceFileManager.clearCache();

    // Clear cached module-specific compilers
    JavaCompilerProvider.getInstance().destory();

    // Cache classpath locations
    for (final Project subModule : project.getSubModules()) {
      if (!(subModule instanceof ModuleProject)) {
        continue;
      }

      SourceFileManager.forModule(((ModuleProject) subModule));
    }

    startOrRestartAnalyzeTimer();
  }

  @NonNull
  @Override
  public CompletionResult complete(final CompletionParams params) {
    final JavaCompilerService compiler = getCompiler(params.getFile());
    if (compiler == null
        || !settings.completionsEnabled()
        || !this.completionProvider.canComplete(params.getFile())) {
      return CompletionResult.EMPTY;
    }

    this.completionProvider.reset(
        compiler, this.settings, this.cachedCompletion, this::updateCachedCompletion);
    return this.completionProvider.complete(params);
  }

  @NonNull
  @Override
  public ReferenceResult findReferences(@NonNull ReferenceParams params) {
    final JavaCompilerService compiler = getCompiler(params.getFile());
    if (!settings.referencesEnabled() || compiler == null) {
      return new ReferenceResult(Collections.emptyList());
    }

    return new ReferenceProvider(compiler).findReferences(params);
  }

  @NonNull
  @Override
  public DefinitionResult findDefinition(@NonNull DefinitionParams params) {
    final JavaCompilerService compiler = getCompiler(params.getFile());
    if (!settings.definitionsEnabled() || compiler == null) {
      return new DefinitionResult(Collections.emptyList());
    }

    return new DefinitionProvider(compiler, getSettings()).findDefinition(params);
  }

  @NonNull
  @Override
  public Range expandSelection(@NonNull ExpandSelectionParams params) {
    final JavaCompilerService compiler = getCompiler(params.getFile());
    if (!settings.smartSelectionsEnabled() || compiler == null) {
      return params.getSelection();
    }

    return new JavaSelectionProvider(compiler).expandSelection(params);
  }

  @NonNull
  @Override
  public SignatureHelp signatureHelp(@NonNull SignatureHelpParams params) {
    final JavaCompilerService compiler = getCompiler(params.getFile());
    if (!settings.signatureHelpEnabled() || compiler == null) {
      return new SignatureHelp(Collections.emptyList(), -1, -1);
    }

    return new SignatureProvider(compiler).signatureHelp(params);
  }

  @NonNull
  @Override
  public DiagnosticResult analyze(@NonNull Path file) {
    if (!DocumentUtils.isJavaFile(file)) {
      return DiagnosticResult.NO_UPDATE;
    }
    
    final JavaCompilerService compiler = getCompiler(file);
    if (!settings.codeAnalysisEnabled() || compiler == null) {
      return DiagnosticResult.NO_UPDATE;
    }

    return this.diagnosticProvider.analyze(file);
  }

  @NonNull
  @Override
  public CodeFormatResult formatCode(FormatCodeParams params) {
    return new CodeFormatProvider(getSettings()).format(params);
  }

  @Override
  public boolean handleFailure(final LSPFailure failure) {
    //noinspection SwitchStatementWithTooFewBranches
    switch (failure.getType()) {
      case COMPLETION:
        if (CancelChecker.isCancelled(failure.getError())) {
          return true;
        }

        JavaCompilerProvider.getInstance().destory();
        return true;
    }

    return false;
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
  public JavaCompilerService getCompiler(Path file) {
    if (!DocumentUtils.isJavaFile(file)) {
      return JavaCompilerService.NO_MODULE_COMPILER;
    }

    final Project root = ProjectManager.INSTANCE.getRootProject();
    if (root == null) {
      return JavaCompilerService.NO_MODULE_COMPILER;
    }

    final ModuleProject module = root.findModuleForFile(file);
    if (module == null) {
      return JavaCompilerService.NO_MODULE_COMPILER;
    }

    return JavaCompilerProvider.get(module);
  }

  private void updateCachedCompletion(CachedCompletion cachedCompletion) {
    Objects.requireNonNull(cachedCompletion);
    this.cachedCompletion = cachedCompletion;
  }

  private void startOrRestartAnalyzeTimer() {

    if (VMUtils.isJvm()) {
      return;
    }

    if (!this.timer.isStarted()) {
      this.timer.start();
    } else {
      this.timer.restart();
    }
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @SuppressWarnings("unused")
  public void onContentChange(@NonNull DocumentChangeEvent event) {
    if (!DocumentUtils.isJavaFile(event.getChangedFile())) {
      return;
    }

    // TODO Find an alternative to efficiently update changeDelta in JavaCompilerService instance
    JavaCompilerService.NO_MODULE_COMPILER.onDocumentChange(event);

    final ModuleProject module = ProjectManager.INSTANCE.findModuleForFile(event.getChangedFile());
    if (module != null) {
      final JavaCompilerService compiler = JavaCompilerProvider.get(module);
      compiler.onDocumentChange(event);
    }

    startOrRestartAnalyzeTimer();
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @SuppressWarnings("unused")
  public void onFileSelected(@NonNull DocumentSelectedEvent event) {
    this.selectedFile = event.getSelectedFile();
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @SuppressWarnings("unused")
  public void onFileOpened(@NonNull DocumentOpenEvent event) {
    this.selectedFile = event.getOpenedFile();
    startOrRestartAnalyzeTimer();
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @SuppressWarnings("unused")
  public void onFileClosed(@NonNull DocumentCloseEvent event) {
    if (this.diagnosticProvider != null) {
      this.diagnosticProvider.clearTimestamp(event.getClosedFile());
    }

    if (FileManager.INSTANCE.getActiveDocumentCount() == 0) {
      this.selectedFile = null;
      this.timer.cancel();
    }
  }

  private void analyzeSelected() {
    if (this.selectedFile == null) {
      return;
    }

    CompletableFuture.supplyAsync(() -> analyze(selectedFile))
        .whenComplete(
            ((diagnostics, throwable) -> {
              if (client != null) {
                client.publishDiagnostics(diagnostics);
              }
            }));
  }
}

File: java/lsp/src/main/java/com/itsaky/androidide/lsp/java/JavaLanguageServer.kt
/*
 *  This file is part of AndroidIDE.
 *
 *  AndroidIDE is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  AndroidIDE is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.itsaky.androidide.lsp.java

import androidx.annotation.RestrictTo
import com.itsaky.androidide.eventbus.events.editor.DocumentChangeEvent
import com.itsaky.androidide.eventbus.events.editor.DocumentCloseEvent
import com.itsaky.androidide.eventbus.events.editor.DocumentOpenEvent
import com.itsaky.androidide.eventbus.events.editor.DocumentSelectedEvent
import com.itsaky.androidide.javac.services.fs.CacheFSInfoSingleton
import com.itsaky.androidide.javac.services.fs.CachingJarFileSystemProvider.clearCache
import com.itsaky.androidide.javac.services.fs.CachingJarFileSystemProvider.clearCachesForPaths
import com.itsaky.androidide.lsp.api.ILanguageClient
import com.itsaky.androidide.lsp.api.ILanguageServer
import com.itsaky.androidide.lsp.api.IServerSettings
import com.itsaky.androidide.lsp.internal.model.CachedCompletion
import com.itsaky.androidide.lsp.java.actions.JavaCodeActionsMenu
import com.itsaky.androidide.lsp.java.compiler.JavaCompilerService
import com.itsaky.androidide.lsp.java.compiler.SourceFileManager
import com.itsaky.androidide.lsp.java.models.JavaServerSettings
import com.itsaky.androidide.lsp.java.providers.CodeFormatProvider
import com.itsaky.androidide.lsp.java.providers.CompletionProvider
import com.itsaky.androidide.lsp.java.providers.DefinitionProvider
import com.itsaky.androidide.lsp.java.providers.JavaDiagnosticProvider
import com.itsaky.androidide.lsp.java.providers.JavaSelectionProvider
import com.itsaky.androidide.lsp.java.providers.ReferenceProvider
import com.itsaky.androidide.lsp.java.providers.SignatureProvider
import com.itsaky.androidide.lsp.java.providers.snippet.JavaSnippetRepository.init
import com.itsaky.androidide.lsp.java.utils.AnalyzeTimer
import com.itsaky.androidide.lsp.java.utils.CancelChecker.Companion.isCancelled
import com.itsaky.androidide.lsp.models.CodeFormatResult
import com.itsaky.androidide.lsp.models.CompletionParams
import com.itsaky.androidide.lsp.models.CompletionResult
import com.itsaky.androidide.lsp.models.DefinitionParams
import com.itsaky.androidide.lsp.models.DefinitionResult
import com.itsaky.androidide.lsp.models.DiagnosticResult
import com.itsaky.androidide.lsp.models.ExpandSelectionParams
import com.itsaky.androidide.lsp.models.FailureType
import com.itsaky.androidide.lsp.models.FormatCodeParams
import com.itsaky.androidide.lsp.models.LSPFailure
import com.itsaky.androidide.lsp.models.ReferenceParams
import com.itsaky.androidide.lsp.models.ReferenceResult
import com.itsaky.androidide.lsp.models.SignatureHelp
import com.itsaky.androidide.lsp.models.SignatureHelpParams
import com.itsaky.androidide.lsp.util.LSPEditorActions
import com.itsaky.androidide.models.Range
import com.itsaky.androidide.projects.FileManager.getActiveDocumentCount
import com.itsaky.androidide.projects.IProjectManager.Companion.getInstance
import com.itsaky.androidide.projects.IWorkspace
import com.itsaky.androidide.projects.ModuleProject
import com.itsaky.androidide.utils.DocumentUtils
import com.itsaky.androidide.utils.VMUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.slf4j.LoggerFactory
import java.nio.file.Path
import java.util.Objects

class JavaLanguageServer : ILanguageServer {

  private val completionProvider: CompletionProvider = CompletionProvider()
  private val diagnosticProvider: JavaDiagnosticProvider?
  override var client: ILanguageClient? = null
    private set

  private var _settings: IServerSettings? = null
  private var selectedFile: Path? = null
  private val timer = AnalyzeTimer { analyzeSelected() }
  private var cachedCompletion: CachedCompletion

  val settings: IServerSettings
    get() {
      return _settings ?: JavaServerSettings.getInstance()
        .also { _settings = it }
    }

  override val serverId: String = SERVER_ID

  companion object {

    const val SERVER_ID = "ide.lsp.java"
    private val log = LoggerFactory.getLogger(JavaLanguageServer::class.java)
  }

  init {
    diagnosticProvider = JavaDiagnosticProvider()
    cachedCompletion = CachedCompletion.EMPTY

    applySettings(JavaServerSettings.getInstance())

    if (!EventBus.getDefault().isRegistered(this)) {
      EventBus.getDefault().register(this)
    }

    init()
  }

  override fun shutdown() {
    JavaCompilerProvider.getInstance().destroy()
    SourceFileManager.clearCache()
    CacheFSInfoSingleton.clearCache()
    clearCache()
    EventBus.getDefault().unregister(this)
    timer.cancel()
  }

  override fun connectClient(client: ILanguageClient?) {
    this.client = client
  }

  override fun applySettings(settings: IServerSettings?) {
    this._settings = settings
  }

  override fun setupWorkspace(workspace: IWorkspace) {
    LSPEditorActions.ensureActionsMenuRegistered(JavaCodeActionsMenu)

    // Once we have workspace initialized
    // Destory the NO_MODULE_COMPILER instance
    JavaCompilerService.NO_MODULE_COMPILER.destroy()

    // Clear cached file managers
    SourceFileManager.clearCache()

    // Clear cached JAR file system for R.jar
    // Using the cached instance will result in completions not being updated for updated resources
    // TODO Clearing caches for JAR files ending with '/R.jar' is probably not a good idea
    //    Maybe this could be improved by using data from the AndroidModule workspace model
    clearCachesForPaths { path: String -> path.endsWith("/R.jar") }

    // Clear cached module-specific compilers
    JavaCompilerProvider.getInstance().destroy()

    // Cache classpath locations
    for (subModule in workspace.getSubProjects()) {
      if (subModule !is ModuleProject || subModule.path == workspace.getRootProject().path) {
        continue
      }
      SourceFileManager.forModule(subModule)
    }
    startOrRestartAnalyzeTimer()
  }

  override fun complete(params: CompletionParams?): CompletionResult {
    val compiler = getCompiler(params!!.file)
    if (!settings.completionsEnabled() || !completionProvider.canComplete(params.file)
    ) {
      return CompletionResult.EMPTY
    }

    if (diagnosticProvider!!.isAnalyzing()) {
      log.warn("Cancelling source code analysis due to completion request")
      diagnosticProvider.cancel()
    }

    completionProvider.reset(
      compiler, settings, cachedCompletion
    ) { cachedCompletion: CachedCompletion ->
      updateCachedCompletion(cachedCompletion)
    }

    return completionProvider.complete(params)
  }

  override suspend fun findReferences(params: ReferenceParams): ReferenceResult {
    val compiler = getCompiler(params.file)
    return if (!settings.referencesEnabled()) {
      ReferenceResult(emptyList())
    } else ReferenceProvider(compiler, params.cancelChecker).findReferences(params)
  }

  override suspend fun findDefinition(params: DefinitionParams): DefinitionResult {
    val compiler = getCompiler(params.file)
    return if (!settings.definitionsEnabled()) {
      DefinitionResult(emptyList())
    } else DefinitionProvider(compiler, settings, params.cancelChecker).findDefinition(params)
  }

  override suspend fun expandSelection(params: ExpandSelectionParams): Range {
    val compiler = getCompiler(params.file)
    return if (!settings.smartSelectionsEnabled()) {
      params.selection
    } else JavaSelectionProvider(compiler).expandSelection(params)
  }

  override suspend fun signatureHelp(params: SignatureHelpParams): SignatureHelp {
    val compiler = getCompiler(params.file)
    return if (!settings.signatureHelpEnabled()) {
      SignatureHelp(emptyList(), -1, -1)
    } else SignatureProvider(compiler, params.cancelChecker).signatureHelp(params)
  }

  override suspend fun analyze(file: Path): DiagnosticResult {
    if (!settings.diagnosticsEnabled() || !DocumentUtils.isJavaFile(file)) {
      return DiagnosticResult.NO_UPDATE
    }

    return if (!settings.codeAnalysisEnabled()) {
      DiagnosticResult.NO_UPDATE
    } else diagnosticProvider!!.analyze(file)
  }

  override fun formatCode(params: FormatCodeParams?): CodeFormatResult {
    return CodeFormatProvider(settings).format(params)
  }

  override fun handleFailure(failure: LSPFailure?): Boolean {
    return when (failure!!.type) {
      FailureType.COMPLETION -> {
        if (isCancelled(failure.error)) {
          return true
        }
        JavaCompilerProvider.getInstance().destroy()
        true
      }
    }
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
  fun getCompiler(file: Path?): JavaCompilerService {
    if (!DocumentUtils.isJavaFile(file)) {
      return JavaCompilerService.NO_MODULE_COMPILER
    }
    val workspace = getInstance().getWorkspace()
      ?: return JavaCompilerService.NO_MODULE_COMPILER
    val module = workspace.findModuleForFile(file!!) ?: return JavaCompilerService.NO_MODULE_COMPILER
    return JavaCompilerProvider.get(module)
  }

  private fun updateCachedCompletion(cachedCompletion: CachedCompletion) {
    Objects.requireNonNull(cachedCompletion)
    this.cachedCompletion = cachedCompletion
  }

  private fun startOrRestartAnalyzeTimer() {
    if (VMUtils.isJvm()) {
      return
    }
    if (!timer.isStarted) {
      timer.start()
    } else {
      timer.restart()
    }
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @Suppress("unused")
  fun onContentChange(event: DocumentChangeEvent) {
    if (!DocumentUtils.isJavaFile(event.changedFile)) {
      return
    }

    // TODO Find an alternative to efficiently update changeDelta in JavaCompilerService instance
    JavaCompilerService.NO_MODULE_COMPILER.onDocumentChange(event)
    val module = getInstance()
      .getWorkspace()?.findModuleForFile(event.changedFile, true)
    if (module != null) {
      val compiler = JavaCompilerProvider.get(module)
      compiler.onDocumentChange(event)
    }
    startOrRestartAnalyzeTimer()
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @Suppress("unused")
  fun onFileSelected(event: DocumentSelectedEvent) {
    selectedFile = event.selectedFile
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @Suppress("unused")
  fun onFileOpened(event: DocumentOpenEvent) {
    selectedFile = event.openedFile
    startOrRestartAnalyzeTimer()
  }

  @Subscribe(threadMode = ThreadMode.ASYNC)
  @Suppress("unused")
  fun onFileClosed(event: DocumentCloseEvent) {
    diagnosticProvider?.clearTimestamp(event.closedFile)

    if (getActiveDocumentCount() == 0) {
      selectedFile = null
      timer.cancel()
    }
  }

  private fun analyzeSelected() {
    if (selectedFile == null || client == null) {
      return
    }

    CoroutineScope(Dispatchers.Default).launch {
      val result = analyze(selectedFile!!)
      withContext(Dispatchers.Main) {
        client?.publishDiagnostics(result)
      }
    }
  }
}